docs/300_OBSERVER.md

```diff
@@ -0,0 +1,76 @@
+# Generic Observer
+An `Observer` evaluates upkeep state from a chain registry and collects upkeeps
+that need to be performed along with data required for comparison and reporting.
+
+Examples can include:
+- polling observer (polls the chain at each block)
+- event trigger observer (logs can trigger observations)
+- scheduled observer (polling occurs on a defined schedule like cron)
+
+## Observer Rules
+- an observer should only be used for observation triggers and timing
+- an observer should be unaware of registry specific types
+- an observer should be chain agnostic
+- an observer should interact with a registry through an interface
+- an observer can define its own dependency interfaces
+- an observer should use a `Coordinator` to track in-flight state
+- all observers used by the same plugin instance should use the same registry
+
+## Filter
+A `Filter` tracks the pending state of upkeeps from the perspective of a single
+node. Once an upkeep is accepted into a report, the filter should indicate
+that upkeep as pending until a completion log is encountered or after a lockout
+window. A `Filter` should be directly paired with an `Observer` as they work in
+tandem to ensure accurate observations. The `Plugin` should never interact 
+directly with a `Filter`.
+
+## Interface
+An `Observer` provides two functions as described in the following interface.
+
+```
+type Observer interface {
+    // Observe provides a list of Eligible upkeep identifiers and the `BlockKey`
+    // they were checked at
+    Observe(context.Context) (BlockKey, []UpkeepIdentifier, error)
+    // MakeReport produces a report from upkeep keys; this is a function on the
+    // Observer because on-chain checks will need to be run as well as cache
+    // hits on the Observer itself.
+    MakeReport(context.Context, []UpkeepKey) ([]byte, int, error)
```

Comment:
intuitively not sure this makes sense on an observer. This seems like a separate task of merging observations that observers from different nodes provide

Reply:
I agree. Instead I kept the `CheckUpkeep` concept and moved the report building to the encoder. Therefore the separation of responsibility is as follows:

Validate and reduce incoming observations: Plugin
Run check upkeep verification: Observer (provides caching) -> Registry
Build report from results: Encoder